src/transformers/models/fast/modeling_fast.py

```diff
@@ -0,0 +1,932 @@
+
+    if reduce:
+        loss_batch = torch.mean(loss_batch)
+
+    return loss_batch
+
+
+def dice_loss_with_masks(input, target, mask, reduce=True):
+    loss_weight = 0.5
+    batch_size = input.size(0)
+    input = torch.sigmoid(input)
+
+    input = input.contiguous().view(batch_size, -1)
+    target = target.contiguous().view(batch_size, -1).float()
+    mask = mask.contiguous().view(batch_size, -1).float()
+
+    input = input * mask
+    target = target * mask
+
+    a = torch.sum(input * target, dim=1)
+    b = torch.sum(input * input, dim=1) + 0.001
+    c = torch.sum(target * target, dim=1) + 0.001
+    d = (2 * a) / (b + c)
+    loss = 1 - d
+
+    loss = loss_weight * loss
+
+    if reduce:
+        loss = torch.mean(loss)
+
+    return loss
+
+
+def ohem_single(score, gt_text, training_mask):
+    pos_num = int(torch.sum(gt_text > 0.5)) - int(torch.sum((gt_text > 0.5) & (training_mask <= 0.5)))
+
+    if pos_num == 0:
+        # selected_mask = gt_text.copy() * 0 # may be not good
+        selected_mask = training_mask
+        selected_mask = selected_mask.view(1, selected_mask.shape[0], selected_mask.shape[1]).float()
+        return selected_mask
+
+    neg_num = int(torch.sum(gt_text <= 0.5))
+    neg_num = int(min(pos_num * 3, neg_num))
+
+    if neg_num == 0:
+        selected_mask = training_mask
+        selected_mask = selected_mask.view(1, selected_mask.shape[0], selected_mask.shape[1]).float()
+        return selected_mask
+
+    neg_score = score[gt_text <= 0.5]
+    neg_score_sorted, _ = torch.sort(-neg_score)
+    threshold = -neg_score_sorted[neg_num - 1]
+
+    selected_mask = ((score >= threshold) | (gt_text > 0.5)) & (training_mask > 0.5)
+    selected_mask = selected_mask.reshape(1, selected_mask.shape[0], selected_mask.shape[1]).float()
+    return selected_mask
+
+
+def ohem_batch(scores, gt_texts, training_masks):
+    selected_masks = []
+    for i in range(scores.shape[0]):
+        selected_masks.append(ohem_single(scores[i, :, :], gt_texts[i, :, :], training_masks[i, :, :]))
+
+    selected_masks = torch.cat(selected_masks, 0).float()
+    return selected_masks
+
+
+def iou_single(a, b, mask, n_class):
+    EPS = 1e-6
+    valid = mask == 1
+    a = a[valid]
+    b = b[valid]
+    miou = []
+    for i in range(n_class):
+        inter = ((a == i) & (b == i)).float()
+        union = ((a == i) | (b == i)).float()
+
+        miou.append(torch.sum(inter) / (torch.sum(union) + EPS))
+    miou = sum(miou) / len(miou)
+    return miou
+
+
+def iou(a, b, mask, n_class=2, reduce=True):
+    batch_size = a.size(0)
+
+    a = a.view(batch_size, -1)
+    b = b.view(batch_size, -1)
+    mask = mask.view(batch_size, -1)
+
+    iou = a.new_zeros((batch_size,), dtype=torch.float32)
+    for i in range(batch_size):
+        iou[i] = iou_single(a[i], b[i], mask[i], n_class)
+
+    if reduce:
+        iou = torch.mean(iou)
+    return iou
+
+
+@dataclass
+class FASTForImageCaptioningOutput(ModelOutput):
```

Comment:
can we not just leverage one of the existing output classes available in the library?

Reply:
Sure, The closest I can find is SemanticSegmenterOutput. But this output class has logits and attention field which current model output does not have . Let me know which other class can be used .